### PR TITLE
Coordinated downing improvements

### PR DESCRIPTION
**Pull request by zbynek001**
contains 

https://github.com/akkadotnet/akka.net/pull/3263
- ClusterSingletonManager should ignore FSM events during shutdown

https://github.com/akka/akka/issues/24048
- Run all CoordinatedShutdown phases also when downing
- add Reason to CoordinatedShutdown

https://github.com/akka/akka/issues/24140
- Allow member to leave a cluster via CoordinatedShutdown.run when MemberStatus is Joining/WeaklyUp/Up.

https://github.com/akka/akka/issues/23449
- ClusterSpec, race between MemberRemoved and MemberExited

### CODE REVIEW

**Comment by Horusiath** [comment]
To be honest I think that #3284 may have covered this PR.

**Comment by zbynek001** [comment]
It might, but I guess that's still far away.
Just wanted to share it when I already migrated it. But it's up to you whether you merge it or not. If not, will just add it to my custom build.

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/CoordinatedShutdownShardingSpec.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        protected override void BeforeTermination()
+        {
+            Shutdown(_sys1);
+            Shutdown(_sys2);
+        }
+
+        /// <summary>
+        /// Using region 2 as it is not shutdown in either test.
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/CoordinatedShutdownShardingSpec.cs]: Good to know (the comment was helpful during my review :p)

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs [hunk 2]
```diff
@@ -379,8 +379,15 @@ private void SetupCoordinatedShutdown()
             var self = Self;
             _coordShutdown.AddTask(CoordinatedShutdown.PhaseClusterShardingShutdownRegion, "region-shutdown", () =>
             {
-                self.Tell(GracefulShutdown.Instance);
-                return _gracefulShutdownProgress.Task;
+                if (Cluster.IsTerminated || Cluster.SelfMember.Status == MemberStatus.Down)
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs]: Looks good

File: src/core/Akka/Actor/CoordinatedShutdown.cs [hunk 3]
```diff
@@ -152,6 +152,76 @@ public static CoordinatedShutdown Get(ActorSystem sys)
...
 
+
+
+        /// <summary>
+        /// Reason for the shutdown, which can be used by tasks in case they need to do
+        /// different things depending on what caused the shutdown. There are some
+        /// predefined reasons, but external libraries applications may also define
+        /// other reasons.
+        /// </summary>
+        public class Reason
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/CoordinatedShutdown.cs]: I like this!

Comment (Horusiath) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: Actually this should be an enum.

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 4]
```diff
@@ -1668,7 +1668,7 @@ public void Welcome(Address joinWith, UniqueAddress from, Gossip gossip)
         public void Leaving(Address address)
         {
             // only try to update if the node is available (in the member ring)
-            if (_latestGossip.Members.Any(m => m.Address.Equals(address) && m.Status == MemberStatus.Up))
+            if (_latestGossip.Members.Any(m => m.Address.Equals(address) && (m.Status == MemberStatus.Joining || m.Status == MemberStatus.WeaklyUp || m.Status == MemberStatus.Up)))
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: Resolves https://github.com/akkadotnet/akka.net/pull/2347/ while we're at it.

File: src/core/Akka/Actor/CoordinatedShutdown.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: Disagree. If it needs properties and data that can be overridden in the future, for whatever reason, you can't have that extensibility with an `Enum`

Comment (Aaronontheweb) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: I.E. the `CustomReason` example used in the spec earlier illustrates one such case for that.

Comment (Horusiath) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: Then you can easily change it into tagged union or just change from enum to class. It's time to stop putting enum-like types into singletons:

1. They blow up the code base.
2. They are hardly maintainable. You need a lot of boilerplate to keep things like equality, hash codes and string formatting in place.
3. They are not safe - you never know if you have covered all switch cases and no tool on earth will help you with that.
4. They are slow and memory-consuming (this don't matter much in this particular case).

If you can model some cases as enums, you should do that. This is not JVM, they won't be optimized automatically.

Comment (Aaronontheweb) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: None of the above is true in this particular case - it's a class that gets used once on shutdown. I don't think these are even true generally.

Comment (zbynek001) [reply src/core/Akka/Actor/CoordinatedShutdown.cs]: I actually started this as an enum, but then I got to the custom reason and I got stuck.
To support custom reasons with enum, we'd need to cast it to another enum extended with the custom reasons and I didn't like that. Honestly, the current approach I also don't like so much

**Pull request merged**